_posts/2021-04-07-publishing-glean-js.markdown

```diff
@@ -0,0 +1,194 @@
+this is how our example can be changed to use conditional exports:
+
+```json
+{
+  "name": "glean",
+  ...
+  "exports": {
+    "./webext": "path/to/entry/point/webext.js",
+    "./node": {
+      "import": "path/to/entry/point/node.js",
+      "require": "path/to/entry/point/node.cjs",
+    },
+    ...
+  }
+  "typesVersions": {
+    "*": {
+      "./webext": [ "path/to/types/definitions/webext.d.ts" ],
+      "./node": [ "path/to/types/definitions/node.d.ts" ],
+      ...
+    }
+  }
+}
+```
+
+The same change is not necessary for the `./webext` entry point, because users building for browsers
+will need to use bundlers such as [Webpack](https://webpack.js.org/) and [Rollup](https://rollupjs.org/guide/en/),
+which have their own implementation of import/require statement resolutions and are able to import
+both ESM and CommonJS modules either [out-of-the-box](https://webpack.js.org/guides/ecma-script-modules/#flagging-modules-as-esm)
+or through [plugins](https://github.com/rollup/plugins/tree/master/packages/commonjs).
+
+> Note that there is also no need to change the `typesVersions` value for `./node` after this change.
+
+## Final considerations
+
+Although the steps in this post look straight forward enough, it took me quite a while to figure out
+the correct way to configure the Glean.js' entry points. I encountered many caveats along the way, such
+as the `typesVersions` workaround I mentioned above, but also:
+
+- In order to support ES6 modules, it is necessary to include the filename and extension in all internal
+package import statements. CommonJS infers the extension and the filename when it is not provided,
+but ES6 doesn't. This get's extra weird in Glean.js' codebase, because Glean.js is in Typescript
+and still all our import statements have the `.js` extension. See more discussion about this on
+[this issue](https://github.com/microsoft/TypeScript/issues/16577#issuecomment-754941937) and
+[our commit with this change](https://github.com/mozilla/glean.js/pull/123/commits/607c9d5285298f7afbc6187d3b2bdd7d0c1f25b3).
+- Webpack below version 5, does not have support for the `exports` field and is not able to import a
+package that defined entry points only using this feature. See the [Webpack 5 release notes](https://webpack.js.org/blog/2020-10-10-webpack-5-release/#major-changes-new-nodejs-ecosystem-features).
+- Other exports conditions such as `browser`, `production` or `development` are mentioned
+in the [Node.js documentation](https://nodejs.org/api/packages.html#packages_conditions_definitions),
+but are ultimately ignored by Node.js. They are used by bundlers such as Webpack and Rollup. The Webpack
+documentation has [a comprehensive list of all the conditions](https://webpack.js.org/guides/package-exports/#conditions)
+you can possibly include in that list, which bundler supports each and whether Node.js supports it too.
+
+Hope this guide is helpful to other people on the internet. Bye! 👋
```

Comment:
Woosh, boy was it ever helpful to me!  I think it was amazing and I learned a ton about the JS implementation of Glean.  Thanks!